src/storage-client/src/controller/command_wals.rs

```diff
@@ -70,63 +70,109 @@ where
+            typed: &TypedCollection<K, V>,
+        ) -> Vec<(K, V, Diff)>
+        where
+            K: mz_stash::Data,
+            V: mz_stash::Data,
+        {
+            let collection = tx
+                .collection::<K, V>(typed.name())
+                .await
+                .expect("named collection must exist");
+            tx.peek(collection).await.expect("peek succeeds")
         }
 
-        // Get all shards we're aware of from stash.
-        let all_shard_data: BTreeMap<_, _> = super::METADATA_COLLECTION
-            .peek_one(&mut self.state.stash)
+        // Get stash metadata.
+        let (metadata, shard_finalization) = self
+            .state
+            .stash
+            .with_transaction(move |tx| {
+                Box::pin(async move {
+                    // Query all collections in parallel.
+                    Ok(futures::join!(
+                        tx_peek(&tx, &super::METADATA_COLLECTION),
+                        tx_peek(&tx, &SHARD_FINALIZATION),
+                    ))
+                })
+            })
             .await
-            .expect("must be able to read from stash")
+            .expect("stash operation succeeds");
+
+        // Partition metadata into the collections we want to have and those we failed to drop.
+        let (in_use_collections, leaked_collections): (Vec<_>, Vec<_>) =
+            metadata.into_iter().partition(|(id, _, diff)| {
+                assert_eq!(
+                    *diff, 1,
+                    "expected METADATA_COLLECTION to contain reconciled state"
+                );
+
+                self.collection(*id).is_ok()
+            });
+
+        // Get all shard IDs
+        let shard_finalization: BTreeSet<_> = shard_finalization
             .into_iter()
-            .map(
-                |(
-                    id,
-                    // TODO(guswynn): produce the schema for each shard.
-                    super::DurableCollectionMetadata {
-                        remap_shard,
-                        data_shard,
-                    },
-                )| { [(id, [remap_shard, Some(data_shard)])] },
-            )
-            .flatten()
+            .map(|(id, _, diff)| {
+                assert_eq!(
+                    diff, 1,
+                    "expected SHARD_FINALIZATION to contain reconciled state"
+                );
+
+                id
+            })
             .collect();
 
-        // Consider shards in use if they are still attached to a collection.
-        let in_use_shards: BTreeSet<_> = all_shard_data
-            .into_iter()
-            .filter_map(|(id, shards)| self.state.collections.get(&id).map(|_| shards.to_vec()))
-            .flatten()
-            .filter_map(|shard| shard)
+        // Collect all shards from in-use collections
+        let in_use_shards: BTreeSet<_> = in_use_collections
+            .iter()
+            .map(|(_, DurableCollectionMetadata { data_shard, .. }, _)| *data_shard)
```

Comment:
Might be worth a `n.b. comment`: we don't record the remap shards here, because those are their own Collection and will be recorded individually, because they're also part of `in_use_collections`.